src/lavinmq/exchange/dedup_ext.cr

```diff
@@ -0,0 +1,67 @@
+module LavinMQ
+  module Deduplication
+    abstract class Cache(T)
+      abstract def contains?(key : T) : Bool
+      abstract def insert(key : T, ttl : UInt32?)
+    end
+
+    class MemoryCache(T) < Cache(T)
+      def initialize(@size : UInt32)
+        @store = Hash(T, Int64?).new
+      end
+
+      def contains?(key : T) : Bool
+        return false unless @store.has_key?(key)
+        ttd = @store[key]
+        return true unless ttd
+        return true if ttd > RoughTime.unix_ms
+        @store.delete(key)
+        false
+      end
+
+      def insert(key : T, ttl : UInt32? = nil)
+        @store.delete(@store.first_key) if @store.size >= @size
+        ttd = RoughTime.unix_ms + ttl if ttl
+        @store[key] = ttd
+      end
+    end
+
+    class Deduper
+      DEFAULT_HEADER_KEY = "x-deduplication-header"
+
+      def initialize(@cache : Cache(AMQ::Protocol::Field), @default_ttl : UInt32? = nil,
+                     @header_key : String? = nil)
+      end
+
+      def add(msg : Message)
+        key = dedup_key(msg)
+        return unless key
+        @cache.insert(key, dedup_ttl(msg))
+      end
+
+      def duplicate?(msg : Message) : Bool
+        key = dedup_key(msg)
+        return false unless key
+        @cache.contains?(key)
+      end
+
+      private def dedup_key(msg)
+        headers = msg.properties.headers
+        return unless headers
+        key = @header_key || DEFAULT_HEADER_KEY
+        headers[key]?
+      end
+
+      private def dedup_ttl(msg) : UInt32?
+        headers = msg.properties.headers
+        def_ttl = @default_ttl
+        return def_ttl unless headers
+        value = headers["x-cache-ttl"]?
+        return def_ttl unless value
+        value = value.try(&.as?(Int32))
+        return def_ttl unless value
+        value.to_u32 || def_ttl
```

Comment:
I do like to keep it as one-lines if possible. But not sure if it's good here since having multple lines makes it easier to read and since it's isolated in it's own method it doesn't conflict with the rest of the code.

Reply:
Yeah, since its isolated maybe its better to have a few extra lines.. and my suggestion has way to many try blocks xd  mostly me just doing exercise 
But.. i do feel like its easier to read if we can avoid changing the value of `value` so many times